### PR TITLE
API: Implement player:{set,get}_fog

### PR DESCRIPTION
**Pull request by kilbith**
Check documentation.

Demo: https://www.youtube.com/watch?v=PmKNEijOv6U

### CODE REVIEW

File: src/script/lua_api/l_object.cpp [hunk 1]
```diff
@@ -1197,6 +1197,40 @@ int ObjectRef::l_get_fov(lua_State *L)
...
+int ObjectRef::l_set_fog(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	ObjectRef *ref = checkObject<ObjectRef>(L, 1);
+	RemotePlayer *player = getplayer(ref);
+	if (player == nullptr)
+		return 0;
+
+	float distance = static_cast<f32>(luaL_checknumber(L, 2));
+	distance = rangelim(distance, 0.f, g_settings->getFloat("viewing_range"));
```
Comment (Desour) [inline src/script/lua_api/l_object.cpp:1210]: Isn't this the server-side setting?

Comment (kilbith) [reply src/script/lua_api/l_object.cpp]: `viewing_range` is a client-side setting.

Comment (Desour) [reply src/script/lua_api/l_object.cpp:1210]: Yes, but you're using this here at the server-side.

Comment (kilbith) [reply src/script/lua_api/l_object.cpp]: There are hierarchical layers of Settings. This one is on the client.

Comment (kilbith) [reply src/script/lua_api/l_object.cpp]: Ah, probably that if I call it from `l_object`, it'll try to get it from the server. Not sure. @sfan5 ?
Or is the hierarchy layer set per-setting name?

Comment (Desour) [reply src/script/lua_api/l_object.cpp:1210]: The only setting hierarchy I know is that for example world settings take preference over the global minetest.conf. (And idk of any client-related layer.)
But this hierarchy isn't relevant here. This code is executed on the server, which doesn't have access to the player's settings.

**Comment by lhofhansl** [comment]
Hmm... Fog is not meant as a game feature, but as means to hide the parts of the world that have not been loaded.

At the very least any custom fog distance must be <= `viewing_range` on the client, and should <= the servers `max_block_send_distance`

**Comment by rubenwardy** [comment]
Could we have near and far distances for fog? What about fog color? Doesn't necessarily need to be in this PR, but implementing the API as a table seems short-sighted

**Comment by kilbith** [comment]
> At the very least any custom fog distance must be <= viewing_range on the client, and should <= the servers max_block_send_distance

~~Further can be done calling [`setWantedRange`](https://github.com/minetest/minetest/blob/1f0d042377ed506ad460067495e5911b2fc2aacc/src/network/serverpackethandler.cpp#L502) to restrict the culler to the custom fog distance.~~

> What about fog color?

It already depends on the sky API.

Need a decision whether the param must a table or not.

**Comment by Desour** [comment]
Found an issue for this: #8142

**Comment by kilbith** [comment]
- `set_fog` now passes a table. Ex.: `player:set_fog { distance = 20 }`
- `get_fog` returns a table.

> Further can be done calling [setWantedRange](https://github.com/minetest/minetest/blob/1f0d042377ed506ad460067495e5911b2fc2aacc/src/network/serverpackethandler.cpp#L502) to restrict the culler to the custom fog distance.

Erratum: quick checking of the value passed to it indicates that the culler already takes the fog distance change into consideration from the moment the viewing range is set from `Camera`.

**Comment by kilbith** [comment]
Added support for colored fog. Ex.: `player:set_fog { distance = 20, color = "#ff0" }`

File: src/client/camera.cpp [hunk 2]
```diff
@@ -597,7 +597,11 @@ void Camera::update(LocalPlayer* player, f32 frametime, f32 tool_reload_ratio)
 
 void Camera::updateViewingRange()
 {
-	f32 viewing_range = g_settings->getFloat("viewing_range");
+	LocalPlayer *player = m_client->getEnv().getLocalPlayer();
+	assert(player);
+
+	PlayerFogSpec fog_spec = player->getFog();
+	f32 viewing_range = fog_spec.distance > 0.f ? fog_spec.distance : g_settings->getFloat("viewing_range");
```
Comment (lhofhansl) [inline src/client/camera.cpp:604]: We should range-limit here as well, right? Client might change the setting after last distance was set from the server.

**Pull request closed without merging**